{% extends "layouts/with_leftbar/base.html" %}
{% load static %}
{% load color_range_for_variants %}
{% csrf_token %}
{% block addon_css %}
    <link rel="stylesheet" href="{% static 'home/css/jquery.dataTables.min.css' %}" type="text/css"/>
    <link rel="stylesheet" href="{% static 'home/css/bootstrap2-toggle.min.css' %}" type="text/css"/>
    <link rel="stylesheet" href="{% static 'home/css/yadcf_bootstrap_version.css' %}" type="text/css"/>
    <link rel="stylesheet" href="{% static 'home/css/select2.css' %}" type="text/css"/>
    <link rel="stylesheet" href="{% static 'home/css/modal.css' %}" type="text/css"/>
    <!--This is for the filter, min max, pagination styling -->
    <link rel="stylesheet" type="text/css"
          href="https://cdn.datatables.net/v/dt/jszip-2.5.0/dt-1.10.16/b-1.5.1/b-html5-1.5.1/datatables.min.css"/>
    <link rel="stylesheet" href="{% static 'home/css/construct_browser.css' %}" type="text/css"/>

    <link rel="stylesheet" href="{% static 'gene/css/gene-detail.css' %}" type="text/css"/>
{% endblock addon_css %}


{% block content %}
    {% autoescape off %}
        <div class="row">
            <div class="col-md-2" style="background-color: rgba(224, 225, 221, 0.3); height: inherit;">
                <p style="font-size: 17px; display: block; margin-bottom: 20px;"><strong>Gene name:</strong> <span style="color: #337ab7">{{gene}} ({{geneID}})</span></p>
                {% include '_filter.html' with transcripts=transcripts consequences=consequences variants=variants %}
            </div>
            <div class="col-md-10" >
                <div class="row">
                    <div class="col-xs-12">
                        <div id="gene-2d-chart" style="height: inherit;"></div>
                    </div>
                </div>
                <div class="gene-table-container">
                    {% include '_buttons.html' %}
                    <div id="structures_scrollable_table">
                        <table class="display compact text-nowrap" id="structures_scrollable">
                            <thead id="headers">
                            <tr class="table-filter-dropdown">
                                <th rowspan="2" style="text-align: left; width:150px">Variant</th>
                                <th rowspan="2">Transcript</th>
                                <!-- <th>cDNA_</br>position</th>
                                <th>CDS_</br>position</th> -->
                                <th rowspan="2">Sequence<br>position</th>
                                <th rowspan="2">Wildtype AA</th>
                                <th rowspan="2">Mutant AA</th>
                                <th rowspan="2">Codons</th>
                                <!-- <th>Strand</th> -->
                                <th rowspan="2">Consequence</th>
                                <!-- <th>Impact</th> -->
                                <th rowspan="2">Mean <br> VEP scores</th>
                                <th colspan="40">VEP scores from different algorithms</th>
                            </tr>
                            <tr>
                                <!-- Score block -->

                                <th>Polyphen2_</br>HDIV</th>
                                <th>Polyphen2_</br>HVAR</th>
                                <th>SIFT4G_</br>converted</th>
                                <th>SIFT_</br>converted</th>

                                <th>BayesDel_</br>addAF</th> 
                                <th>BayesDel_</br>noAF</th>
                                <th>CADD_</br>raw</th>
                                <th>ClinPred</th>
                                <th>DANN</th>

                                <th>DEOGEN2</th>
                                <th>Eigen_PC_</br>raw_coding</th>
                                <th>Eigen_raw_</br>coding</th>
                                <th>FATHMM_</br>converted</th>
                                <th>GERP_RS</th>

                                <th>GM12878_</br>fitCons</th>
                                <th>Geno</br>Canyon</th>
                                <th>H1_hESC_</br>fitCons</th>
                                <th>HUVEC_</br>fitCons</th>
                                <th>LIST_S2</th>

                                <th>LRT_</br>converted</th>
                                <th>M_CAP</th>
                                <th>MPC</th>
                                <th>MVP</th>
                                <th>MetaLR</th>

                                <th>MetaRNN</th>
                                <th>MetaSVM</th>
                                <th>MutPred</th>
                                <th>Mutation</br>Assessor</th>
                                <th>Mutation</br>Taster_</br>converted</th>

                                <th>PROVEAN_</br>converted</th>
                                <!-- <th>Polyphen2_</br>HDIV</th>
                                <th>Polyphen2_</br>HVAR</th> -->
                                <th>PrimateAI</th>
                                <th>REVEL</th>

                                <!-- <th>SIFT4G_</br>converted</th>
                                <th>SIFT_</br>converted</th> -->
                                <th>SiPhy_</br>29way_</br>logOdds</th>
                                <th>VEST4</th>
                                <th>bStatistic</br>converted</th>

                                <th>Fathmm_</br>MKL_coding</th>
                                <th>Fathmm_XF</br>_coding</th>
                                <th>Integrated_</br>fitCons</th>
                                <th>PhastCons</br>30way_</br>mammalian</th>
                                <th>Phylo_</br>P30way_</br>mammalian</th>
                            </tr>

                            </thead>
                            <tbody id='structures_scrollable_body'>
                            {% for value in array %}
                                <tr>
                                    <td class="align-left"><a href="#" class="variants-table-row"
                                                              data-remote-url="{% url 'variant:get-genebass-tables' %}"
                                                              data-variant-marker="{{ value.0 }}" data-toggle="modal"
                                                              data-target="#genebassModal">{{ value.0 }}</a></td>
                                    <td class="align-left">{{ value.1 }}</td>
                                    <!-- <td class="align-left">{{ value.3 }}</td>
                                    <td class="align-left">{{ value.4 }}</td> -->
                                    <td class="align-left">{{ value.5 }}</td>
                                    <td class="align-left">{{ value.6|slice:":1" }}</td>
                                    <td class="align-left">{{ value.6|slice:"2:3" }}</td>
                                    <td class="align-left">{{ value.7 }}</td>
                                    <!-- <td class="align-left">{{ value.9 }}</td> -->
                                    <td class="align-left">{{ value.2 }}</td>
                                    <!-- <td class="align-left">{{ value.8 }}</td> -->
                                    <td>{{ value|last }}</td>

                                    {% render_variant_values_in_color_ranges value.36 %} <!-- Polyphen2 HDIV-->
                                    {% render_variant_values_in_color_ranges value.37 %} <!-- Polyphen2 HVAR-->
                                    {% render_variant_values_in_color_ranges value.40 %} <!-- SIFT4G converted-->
                                    {% render_variant_values_in_color_ranges value.41 %} <!-- SIFT converted-->

                                    {% render_variant_values_in_color_ranges value.10 %}
                                    {% render_variant_values_in_color_ranges value.11 %}
                                    {% render_variant_values_in_color_ranges value.12 %}
                                    {% render_variant_values_in_color_ranges value.13 %}
                                    {% render_variant_values_in_color_ranges value.14 %}
                                    {% render_variant_values_in_color_ranges value.15 %}
                                    {% render_variant_values_in_color_ranges value.16 %}
                                    {% render_variant_values_in_color_ranges value.17 %}
                                    {% render_variant_values_in_color_ranges value.18 %}
                                    {% render_variant_values_in_color_ranges value.19 %}
                                    {% render_variant_values_in_color_ranges value.20 %}
                                    {% render_variant_values_in_color_ranges value.21 %}
                                    {% render_variant_values_in_color_ranges value.22 %}
                                    {% render_variant_values_in_color_ranges value.23 %}
                                    {% render_variant_values_in_color_ranges value.24 %}
                                    {% render_variant_values_in_color_ranges value.25 %}
                                    {% render_variant_values_in_color_ranges value.26 %}
                                    {% render_variant_values_in_color_ranges value.27 %}
                                    {% render_variant_values_in_color_ranges value.28 %}
                                    {% render_variant_values_in_color_ranges value.29 %}
                                    {% render_variant_values_in_color_ranges value.30 %}
                                    {% render_variant_values_in_color_ranges value.31 %}
                                    {% render_variant_values_in_color_ranges value.32 %}
                                    {% render_variant_values_in_color_ranges value.33 %}
                                    {% render_variant_values_in_color_ranges value.34 %}
                                    {% render_variant_values_in_color_ranges value.35 %}
                                    <!-- {% render_variant_values_in_color_ranges value.36 %}
                                    {% render_variant_values_in_color_ranges value.37 %} -->
                                    {% render_variant_values_in_color_ranges value.38 %}
                                    {% render_variant_values_in_color_ranges value.39 %}
                                    <!-- {% render_variant_values_in_color_ranges value.40 %}
                                    {% render_variant_values_in_color_ranges value.41 %} -->
                                    {% render_variant_values_in_color_ranges value.42 %}
                                    {% render_variant_values_in_color_ranges value.43 %}
                                    {% render_variant_values_in_color_ranges value.44 %}
                                    {% render_variant_values_in_color_ranges value.45 %}
                                    {% render_variant_values_in_color_ranges value.46 %}
                                    {% render_variant_values_in_color_ranges value.47 %}
                                    {% render_variant_values_in_color_ranges value.48 %}
                                    {% render_variant_values_in_color_ranges value.49 %}
                                </tr>
                            {% endfor %}
                            </tbody>
                        </table>
                        <br/>
                    </div>

                <div class="col-sm-12" id="gene-3d-chart" style="padding: 0px; margin-bottom: 30px;">
                    {% include "_primary_seq.html" %}
                    <div class="col-sm-3" style="padding: 0px;"> 
                        <p style="font-size:16px; padding-left:0; margin-left:0; color:#337ab7;">3D structure:</p>
                        <div id="filter_3d">
                            <div id="filter_3d_source">
                                <p>Source:</p>
                                <select id="sourceSelect" name="sourceSelect">
                                    <option value="pdb">PDB</option>
                                    <option value="af2">AlphaFold2</option>
                                </select>
                            </div>
                            <div id="filter_3d_color">
                                <p>Color:</p>
                                <select id="colorSelect" name="color">
                                    <option value="spectrum" selected="selected">Spectrum</option>
                                    <option value="grey">Grey</option>
                                </select>
                            </div>
                            <div id="filter_3d_variant">
                                <p>Variant:</p>
                                <select id="visibilitySelect" name="variant">
                                    <option value="hide">Hide</option>
                                    <option value="show">Show</option>
                                </select>
                            </div>
                        </div>
                    </div>
                    <div id="structure_3D" class="col-sm-9" > 
                        {% include '_3d-view.html' %}
                    </div>
                    
                </div>
                
            </div>
        </div>
        <!-- The modal -->
        <div class="modal fade" id="genebassModal" tabindex="-1" role="dialog" aria-labelledby="modalLabel"
             aria-hidden="true">
            <div class="modal-dialog" role="document">
                <div class="modal-content" id="genebass-variant-list">
                    Data will be loaded here by jQuery
                </div>
            </div>
        </div>

    {% endautoescape %}
{% endblock content %}

{% block addon_js %}
    <script src="{% static 'home/js/datatables.min.js' %}"></script>
    <script src="{% static 'home/js/bootstrap2-toggle.min.js' %}"></script>
    <script src="{% static 'home/js/jquery.dataTables.yadcf.js' %}"></script>
    <script src="{% static 'home/js/dataTables.buttons.min.js' %}"></script>
    <script src="{% static 'home/js/select2.js' %}"></script>
    <script src="{% static 'home/js/grayscale.js' %}"></script>
    <script src="{% static 'home/js/gpcrdb.js' %}"></script>
    <script src="{% static 'home/js/signprot-multitabtable.js' %}"></script>
    <script src='https://cdn.plot.ly/plotly-2.20.0.min.js'></script>

    <!-- Script for ploting the 2D gene viewer -->
    <script>
        let list_variants = [];
        let list_transcripts = [];
        let list_vep_variants_consequence = [];
        let list_vep_variants_pos = [];
        let list_vep_variants_mean_vep_score = [];
        {% for element in array %}
            list_variants.push("{{element.0}}");
            list_transcripts.push("{{element.1}}");
            list_vep_variants_consequence.push("{{element.2}}");
            list_vep_variants_pos.push({{element.5}});
            list_vep_variants_mean_vep_score.push({{element|last}});
        {% endfor %}

        let maxValue = list_vep_variants_pos[0]; // Initialize maxValue with the first element
        let minValue = list_vep_variants_pos[0]; // Initialize minValue with the first element
        let maxIndex = 0; // Initialize maxIndex with 0
        let minIndex = 0; // Initialize minIndex with 0

        for (let i = 1; i < list_vep_variants_pos.length; i++) {
        if (list_vep_variants_pos[i] > maxValue) {
            maxValue = list_vep_variants_pos[i];
            maxIndex = i;
        }
        if (list_vep_variants_pos[i] < minValue) {
            minValue = list_vep_variants_pos[i];
            minIndex = i;
        }
        }

        // console.log("Number of variants:" + list_variants.length);
        // console.log("Number of positions:" + list_vep_variants_pos.length);
        // console.log("Maximum value:" + maxValue);
        // console.log("Index of maximum value:" + maxIndex);
        // console.log("Minimum value:" + minValue);
        // console.log("Index of minimum value:" + minIndex);
        // console.log("Variant at min index "+ list_variants[minIndex])
        // console.log("Transcript at min index "+ list_transcripts[minIndex])
        // console.log("Variant at max index "+ list_variants[maxIndex])
        // console.log("Transcript at max index "+ list_transcripts[maxIndex])

        list_colors = [];
        for (let i = 0; i < list_vep_variants_consequence.length; i++) {
            if (list_vep_variants_consequence[i].includes('Missense')) {
                list_colors.push('red');
            } else if (list_vep_variants_consequence[i].includes('Synonymous')) {
                list_colors.push('#8338EC');
            } else if (list_vep_variants_consequence[i].includes('Frameshift')) {
                list_colors.push('green');
            } else if (list_vep_variants_consequence[i].includes('Stop_gained')) {
                list_colors.push('#3A86FF')
            } else if (list_vep_variants_consequence[i].includes('Stop_lost')) {
                list_colors.push('orange')
            } else {
                list_colors.push('black');
            }
        }

        max_xaxis = Math.max(...list_vep_variants_pos);

        let trace1 = {
            x: list_vep_variants_pos,
            y: list_vep_variants_mean_vep_score,
            mode: 'markers',
            marker: {
            size: 7,
            color: list_colors,
            },
            type: 'scatter',
            text: list_variants.map((variant, index) => `Variant: ${variant}`),
            transcript: list_transcripts,
            cliponaxis: false,
        };


        let data = [trace1];

        let layout = {
            xaxis: {
                title: {
                    text: 'Position on protein sequence',
                    font: {
                        size: 14,
                        color: '#337ab7'
                    }
                },
                range: [-2, parseInt(max_xaxis * 1.05)],
                showline: true, // Show x-axis line
                showgrid: true, // Show x-axis grid lines
                zeroline: false, // Show the x-axis zero line
            },
            yaxis: {
                title: {
                    text: 'Mean of VEP scores',
                    font: {
                        size: 14,
                        color: '#337ab7'
                    }
                },
                range: [0.01, 1.05],
                showline: true, // Show y-axis line
                showgrid: true, // Show y-axis grid lines
                zeroline: false, // Show the y-axis zero line
            },
            margin: {
                t: 10, // Increase the top margin for the title and plot container
                b: 60, // Bottom margin for the x-axis labels and title
                l: 60, // Left margin for the y-axis labels and title
                r: 40, // Right margin
            }
        };

        Plotly.newPlot('gene-2d-chart', data, layout);

    </script>

    <!-- Script for showing datatable of variants and VEP scores  -->
    <script>
        $(document).ready(function () {
            let structures_scrollable = $("#structures_scrollable")
            structures_scrollable.show();
            let oTable1 = structures_scrollable.DataTable({
                deferRender: true,
                scrollY: true,
                scrollX: true,
                scrollCollapse: true,
                scroller: true,
                paging: true,
                bSortCellsTop: false, //prevent sort arrows going on bottom row
                aaSorting: [],
                autoWidth: true,
                bInfo: true,
                //order: [[7, "asc"]],
                pagingType: 'full_numbers',
                lengthMenu: [
                    [10, 25, 50, 100, 200, -1],
                    [10, 25, 50, 100, 200, 'All'],
                ],
            })
            $('#lower_mean').change(function () {
                    let filter_params = get_filter_params();
                    update_data(filter_params)
                }
            );
            $('#upper_mean').change(function () {
                    let filter_params = get_filter_params();
                    update_data(filter_params)
                }
            );
            $('#variant-select').change(function () {
                    let filter_params = get_filter_params();
                    update_data(filter_params)
                }
            );
            $('#transcript-select').change(function () {

                    let filter_params = get_filter_params();
                    update_data(filter_params)
                }
            );
            $('#consequence-select').change(function () {

                    let filter_params = get_filter_params();
                    update_data(filter_params)
                }
            );
        })

        // Get the filter values
        function get_filter_params() {
            
            let mean_vep_score_lower = $('#lower_mean').val();
            let mean_vep_score_upper = $('#upper_mean').val();
            let transcript = $('#transcript-select').val();
            let variant = $('#variant-select').val();
            let consequence = $('#consequence-select').val();

            return {
                mean_vep_score_lower: mean_vep_score_lower,
                mean_vep_score_upper: mean_vep_score_upper,
                transcript: transcript,
                variant: variant,
                consequence: consequence,
            };
        }

        function update_data(filter_values) {
            // alert("update data function is called")
            redraw_2d_chart(filter_values)
            redraw_vep_variants_datatable(filter_values)
        }

        function redraw_2d_chart(filter_values) {
            // Test data
            let filtered_list_vep_variants_pos = [];
            let filtered_list_vep_variants_mean_vep_score = [];
            let filtered_list_vep_variants_consequence = [];
            let filtered_list_colors = [];

            for (let i = 0; i < list_vep_variants_pos.length; i++) {
                if (condition_check(list_vep_variants_mean_vep_score[i], list_variants[i], list_vep_variants_consequence[i], list_transcripts[i], filter_values)) {
                    filtered_list_vep_variants_pos.push(list_vep_variants_pos[i]);
                    filtered_list_vep_variants_mean_vep_score.push(list_vep_variants_mean_vep_score[i]);
                    filtered_list_vep_variants_consequence.push(list_vep_variants_consequence[i]);
                    filtered_list_colors.push(list_colors[i]);
                }
            }

            let trace1 = {
                x: filtered_list_vep_variants_pos,
                y: filtered_list_vep_variants_mean_vep_score,
                mode: 'markers',
                marker: {
                size: 7,
                color: list_colors,
                // hovertemplate: 'Mean VEP score: %{y}<br>Position: %{x}<br>Variant: %{text}<br>Transcript: %{transcript}',
                },
                type: 'scatter',
                text: list_variants.map((variant, index) => `Variant: ${variant}`),
                transcript: list_transcripts,
            };

            let data = [trace1];

            let layout = {
            xaxis: {
                title: {
                    text: 'Position on protein sequence',
                    font: {
                        size: 14,
                        color: '#337ab7'
                    }
                },
                range: [-2, parseInt(max_xaxis * 1.05)],
                showline: true, // Show x-axis line
                showgrid: true, // Show x-axis grid lines
                zeroline: false, // Show the x-axis zero line
                
            },
            yaxis: {
                title: {
                    text: 'Mean of VEP scores',
                    font: {
                        size: 14,
                        color: '#337ab7'
                    }
                },
                range: [0.01, 1.05],
                showline: true, // Show y-axis line
                showgrid: true, // Show y-axis grid lines
                zeroline: false, // Show the y-axis zero line
                
            },
            margin: {
                t: 10, // Increase the top margin for the title and plot container
                b: 60, // Bottom margin for the x-axis labels and title
                l: 60, // Left margin for the y-axis labels and title
                r: 40, // Right margin
            }
            }
            Plotly.newPlot('gene-2d-chart', data, layout);
        }

        // this function returns true if a variant, along with its transcript, consequence and vep score pass all the 4 filters
        function condition_check(vep_score, transcript, variant, consequence, filter_values) {
            let vep_score_check = true;
            if ((filter_values.mean_vep_score_lower !== null || filter_values.mean_vep_score_upper !== null) &&
                (vep_score < filter_values.mean_vep_score_lower || vep_score > filter_values.mean_vep_score_upper)) {
                vep_score_check = false;
            }
            let variant_check = true;
            if (filter_values.variant.length !== 0) {
                if (filter_values.variant.indexOf(variant) === -1) {
                    variant_check = false;
                }
            }
            let consequence_check = true;
            if (filter_values.consequence.length !== 0) {
                const intersection = consequence.split(",").filter(element => filter_values.consequence.includes(element));
                if (intersection.length === 0) {
                    consequence_check = false;
                }
            }
            let transcript_check = true;
            if (filter_values.transcript.length !== 0) {
                if (filter_values.transcript.indexOf(transcript) === -1) {
                    transcript_check = false;
                }
            }

            return vep_score_check && variant_check && consequence_check && transcript_check;
        }

        function redraw_vep_variants_datatable(filter_values) {
            // alert("redraw_vep_variants_datatable is called, filter_values.transcript.length = "+filter_values.transcript.length)
            let vep_variants_table = $("#structures_scrollable").DataTable();
            //filter by transcript
            let transcript_search_terms = []
            for (let i = 0; i < filter_values.transcript.length; i++) {
                transcript_search_terms.push("^" + filter_values.transcript[i] + "$")
            }
            vep_variants_table.column(1).search(transcript_search_terms.join('|'), true, false, true).draw();

            //filter by variants
            let variant_search_terms = []
            for (let i = 0; i < filter_values.variant.length; i++) {
                variant_search_terms.push("^" + filter_values.variant[i] + "$")
            }
            vep_variants_table.column(0).search(variant_search_terms.join('|'), true, false, true).draw();

            //filter by consequence
            let consequence_search_terms = []
            for (let i = 0; i < filter_values.consequence.length; i++) {
                consequence_search_terms.push(filter_values.consequence[i])
            }
            vep_variants_table.column(8).search(consequence_search_terms.join('|'), true, false, true).draw();
        }

        $(document).ready(function () {
            let minEl = $('#mean_vep_score')
            $.fn.dataTable.ext.search.push(function (settings, data, dataIndex) {
                let min = parseFloat(minEl.val()) || 0.0;
                let max = 1;
                let mean_vep_score = parseFloat(data[10]) || 0; // use data for the mean_vep_score column
                // console.log(min, max, mean_vep_score)

                if (
                    (isNaN(min) && isNaN(max)) ||
                    (isNaN(min) && mean_vep_score <= max) ||
                    (min <= mean_vep_score && isNaN(max)) ||
                    (min <= mean_vep_score && mean_vep_score <= max)
                ) {
                    return true;
                }

                return false;
            });

            let table = $('#structures_scrollable').DataTable();

            // Changes to the inputs will trigger a redraw to update the table
            minEl.change(function () {
                // console.log(minEl.val())
                table.draw();
            });
        })

    </script>

    <script>
        $(document).ready(function () {
            $('.variants-table-row').click(function (e) {
                e.preventDefault();
                let variant_marker = $(this).data('variant-marker');
                let url = $(this).data('remote-url');

                $("#genebass-variant-list").html('<span class="glyphicon glyphicon-refresh glyphicon-refresh-animate"></span> Genebass variants are loading ...')

                $.ajax(
                    {
                        url: url,
                        type: "GET",
                        data: {
                            'variant_marker': variant_marker,
                        },
                        success: function (response) {
                            $("#genebass-variant-list").html(response)
                            // console.log(response)
                            redraw_genebass_modal()
                        }
                    }
                )
            })
        })

        function redraw_genebass_modal() {
            let genebass_table = $("#genebass-tables")
            genebass_table.show()
            let genebass_variant_table = genebass_table.DataTable({
                searchPanes: {
                    layout: 'columns-1'
                },
                dom: '<"dtsp-dataTable"frtip>',
                deferRender: true,
                scrollY: true,
                scrollX: true,
                scrollCollapse: true,
                scroller: true,
                paging: true,
                bSortCellsTop: false, //prevent sort arrows going on bottom row
                aaSorting: [],
                autoWidth: true,
                bInfo: true,
                //order: [[7, "asc"]],
                pagingType: 'full_numbers',
                lengthMenu: [
                    [10, 25, 50, 100, 200, -1],
                    [10, 25, 50, 100, 200, 'All'],
                ],
            })

            let column_filters = [];

            // Phenocode
            column_filters = column_filters.concat(createYADCFfilters(3, 1, "multi_select", "select2", "Select", false, null, null));

            yadcf.init(genebass_variant_table.draw(), column_filters, {
                cumulative_filtering: false
            });

            gray_scale_table(genebass_table);

            $("#reset").click(function () {
                yadcf.exResetAllFilters(genebass_variant_table);
            });
        }
    </script>

    <!-- Script for filtering options of 3d-structure  -->
    <script>
        const colorSelect = document.getElementById('colorSelect');
        const structureDisplay = document.getElementById('structure_display');

        structureDisplay.dataset.style = 'cartoon:color=blue';

        colorSelect.addEventListener('change',  (event) => {
            const selectedColor = event.target.value;
            if (selectedColor === 'grey') {
                alert("we choose grey-------");
                structureDisplay.dataset.style = 'cartoon:color=grey';
                alert("color changed to: "+structureDisplay.dataset.style);
            } else {
                alert("we choose spectrum");
                structureDisplay.dataset.style = 'cartoon:color=spectrum';
            }
        });

        const visibilitySelect = document.getElementById('visibilitySelect');
        visibilitySelect.addEventListener('change',  (event) => {
            const selectedvisibility = event.target.value;
            if (selectedvisibility === 'hide') {
                alert("we choose hide");
            } else {
                alert("we choose show");
            }
        });
        
        const sourceSelect = document.getElementById('sourceSelect');
        sourceSelect.addEventListener('change',  (event) => {
            const sourceSelect = event.target.value;
            if (sourceSelect === 'pdb') {
                alert("we choose pdb");
            } else {
                alert("we choose af2");
            }
        });
        </script>

    
{% endblock addon_js %}